Implement persistent foreground service to keep calls active in background, with notification controls for managing the call

tareko opened the pull request:
This PR is in response to Issue #1688 . In summary, if PIP is not available, the call ends and this should not happen. I used ChatGPT Codex, Claude, and localAI via Qwen3-coder to generate the code in this PR. Within my personal programming abilities, I still believe the code is elegant. I've used the test app for several calls with no issues on a P6P with latest android.

This PR implements two key functions:
1. It allows the call to remain active even if PIP is not enabled and it is in the background
2. It creates a persistent notification that:
   a. Allows users to return to the call
   b. Allows users to end an active call directly from the persistent notification without needing to return to the CallActivity

## Features
- Add CallForegroundService with persistent notification
- Support calls in background without requiring picture-in-picture mode
- Add "Return to call" and "End call" action buttons to CallForegroundService notification with corresponding PendingIntent
- Handle proper foreground service types for microphone/camera permissions
- Add notification permission and fallback messaging.
- Add EndCallReceiver to handle end call broadcasts from notification action
- Use existing ic_baseline_close_24 drawable for end call action icon
- Register broadcast receiver in CallActivity to handle end call requests from notification using ReceiverFlag.NotExported for Android 14+ compatibility
- Add proper cleanup flow: notification action → EndCallReceiver → CallActivity → proper hangup sequence
- Track intentional call leaving to prevent unwanted service restarts
- Release proximity sensor lock properly during notification-triggered hangup
- Add diagnostic logging throughout the end call flow for debugging

The implementation follows Android best practices:
- Uses NotExported receiver flag for internal app-only broadcasts
- Properly unregisters receivers in onDestroy to prevent leaks
- Uses immutable PendingIntents for security
- Maintains proper state management during call termination

### 🏁 Checklist

- [ ] ⛑️ Tests (unit and/or integration) are included or not needed
- [ ] 🔖 Capability is checked or not needed 
- [x] 🔙 Backport requests are created or not needed: `/backport to stable-xx.x`
- [ ] 📅 Milestone is set
- [ ] 🌸 PR title is meaningful (if it should be in the changelog: is it meaningful to users?)

Commit:
Implement persistent foreground service to keep calls active in background, with notification controls for managing the call

- Add CallForegroundService with persistent notification
- Support calls in background without requiring picture-in-picture mode
- Add "Return to call" and "End call" action buttons to CallForegroundService notification with corresponding PendingIntent
- Handle proper foreground service types for microphone/camera permissions
- Add notification permission and fallback messaging.
- Add EndCallReceiver to handle end call broadcasts from notification action
- Use existing ic_baseline_close_24 drawable for end call action icon
- Register broadcast receiver in CallActivity to handle end call requests from notification using ReceiverFlag.NotExported for Android 14+ compatibility
- Add proper cleanup flow: notification action → EndCallReceiver → CallActivity → proper hangup sequence
- Track intentional call leaving to prevent unwanted service restarts
- Release proximity sensor lock properly during notification-triggered hangup
- Add diagnostic logging throughout the end call flow for debugging

The implementation follows Android best practices:
- Uses NotExported receiver flag for internal app-only broadcasts
- Properly unregisters receivers in onDestroy to prevent leaks
- Uses immutable PendingIntents for security
- Maintains proper state management during call termination

Signed-off-by: Tarek Loubani <tarek@tarek.org>

Commit:
Make logging of warnings and errors more consistent with repository style

Signed-off-by: Tarek Loubani <tarek@tarek.org>

sowjanyakch commented:
Thank you for the PR @tareko
We will review your PR soon.

mahibi commented on app/src/main/java/com/nextcloud/talk/activities/CallActivity.kt:
```diff
@@ -315,12 +322,16 @@ class CallActivity : CallBaseActivity() {
             if (java.lang.Boolean.TRUE == audioPermission) {
                 Log.d(TAG, "Microphone permission was granted")
             } else {
+                Log.d(TAG, "Microphone permission was denied")
```
"...was denied" logs are misleading here.
While the "...was granted" indicate that that some permission is already available, the else branch will just add strings to the dialog that tell why permission xy should be granted.
So the logs should better be "... permission is not yet available and will be asked for"

tareko replied:
cleaned up

mahibi commented on app/src/main/java/com/nextcloud/talk/services/CallForegroundService.kt:
```diff
@@ -57,6 +59,26 @@ class CallForegroundService : Service() {
+        
+        // DIAGNOSTIC: Logging icon resource availability
+        Log.d("CallForegroundService", "Creating end call action - checking icon resources")
+        Log.d("CallForegroundService", "Using ic_baseline_close_24 instead of non-existent ic_close_white_24px")
```
these diagnostic logs/comments about the icon should be deleted?

mahibi reviewed (changes requested):
Thank you @tareko 
Also thanks to let us know that AI was used.

The code looks okay so far from my side.
I tested and it works quite fine.

But i had one situation that is a blocker (could not reproduce it yet):
I think i pressed the hangup button in the app.
The app froze, but audio connection was oneside from web to android.
Even when i killed the app and also the notification was not shown anymore, the onesided audio connection continued.
We will have to test and analyze how this happens.

I might not find enough time in the next weeks to analyze and test appropriately.
So @sowjanyakch and @rapterjet2004 , please also review and test.

@tareko Could you also please fix the warnings in
https://app.codacy.com/gh/nextcloud/talk-android/pull-requests/5660/issues
?

tareko commented on .vscode/settings.json:
```diff
@@ -0,0 +1,3 @@
+{
+    "java.configuration.updateBuildConfiguration": "interactive"
+}
```
Apologies for the delay still. I had to travel suddenly and urgently and will get back to this in 2 or so weeks once things settle. Thank you.

Commit:
Remove .vscode and add to .gitignore

Signed-off-by: Tarek Loubani <tarek@tarek.org>

Commit:
Remove unnecessary logging about icon

Signed-off-by: Tarek Loubani <tarek@tarek.org>

Commit:
Clean up microphone permission language to be more clear

Signed-off-by: Tarek Loubani <tarek@tarek.org>

Commit:
Move endCallFromNotificationReceiver receiver up above companion object

Signed-off-by: Tarek Loubani <tarek@tarek.org>

Commit:
Fix typo to include whole directory

Signed-off-by: Tarek Loubani <tarek@tarek.org>

Commit:
Incorporate refactor from PR #5957 by @rapterjet2004

Signed-off-by: Tarek Loubani <tarek@tarek.org>

Commit:
Fix problem where call does not correctly get switched to PIP if you do a rapid gesture switch back.

For the quick-switch gesture and the recents-to-chat path, the OS starts animating the transition before onPause() fires. By the time onPause()
runs, the window has already been moved off-screen by the gesture animation, so enterPictureInPictureMode() silently fails — Android requires the
window to still be visible.

Why onTopResumedActivityChanged(false) fixes it:
This callback fires when any other activity (including ChatActivity in the same app) takes the "top resumed" slot. Critically, it fires before
onPause() and before any transition animation begins — the window is still fully on-screen. enterPictureInPictureMode() succeeds at this point.

Why back-button worked but this didn't:
Back button goes through OnBackPressedCallback.handleOnBackPressed() synchronously, which calls enterPipMode() before any transition, not in a
lifecycle callback. onTopResumedActivityChanged puts the task-switch path on the same footing.

API compatibility: On API 26–28, onTopResumedActivityChanged is never called by the system (it didn't exist in Activity before API 29), so
onPause() remains the fallback. Older devices primarily use button navigation and won't have the gesture quick-switch anyway.

Signed-off-by: Tarek Loubani <tarek@tarek.org>

Commit:
Fix conversation state race condition when navigating away from chat

Use observeForever for leaveRoom observer so cleanup runs even when
activity is paused. Move ApplicationWideCurrentRoomHolder.clear() into
the leave success callback to avoid premature state clearing. Guard
against double leaveRoom calls with isLeavingRoom flag.

Signed-off-by: Tarek Loubani <tarek@tarek.org>

Commit:
Fix call stability when backgrounding and add PIP/lifecycle diagnostic logging

- Prevent spurious roomJoined events from re-running performCall() when
  already IN_CONVERSATION, fixing call reconnection when ChatActivity
  resumes behind PIP or task switch
- Remove setAutoEnterEnabled(true) which conflicts with manual
  enterPictureInPictureMode() calls causing invisible PIP windows
- Set aspect ratio in initial PIP params (onCreate) so PIP params are
  always valid
- Add isInPipMode guard to onUserLeaveHint to prevent redundant PIP
  entry attempts
- Add diagnostic logging to CallBaseActivity lifecycle methods and
  CallActivity PIP/call state transitions
- Add unit tests documenting PIP race conditions and leaveRoom
  lifecycle behavior

Signed-off-by: Tarek Loubani <tarek@tarek.org>

Commit:
Rewrite PIP entry to follow Android docs: auto-enter + onTopResumedActivityChanged fallback

The previous approach had three competing PIP entry mechanisms on API 31+
(auto-enter, onTopResumedActivityChanged, onPause) that raced against each
other, and onTopResumedActivityChanged toggled setAutoEnterEnabled off/on
which broke smooth transitions.

New layered approach per the Android PIP documentation:
- API 31+: setAutoEnterEnabled(true) as primary for home/recents gestures
- API 29+: onTopResumedActivityChanged as fallback (fires while window is
  still visible, catches quick-switch gestures auto-enter misses)
- API 26-30: onUserLeaveHint for home/recents, onPause fallback for 26-28
- All APIs: OnBackPressedCallback for back gesture (only manual entry point)

Key fix: onTopResumedActivityChanged no longer disables auto-enter. It checks
isInPictureInPictureMode() so if auto-enter already handled it, the manual
call is skipped. No races, no toggling.

Also removes shouldFinishOnStop, pipFallbackHandler, topResumedLostTime, and
onCreateTime which were artifacts of the old racing approach.

Signed-off-by: Tarek Loubani <tarek@tarek.org>

Commit:
Keep call alive when task-switching away from CallActivity

Remove excludeFromRecents=true from CallActivity manifest entry. This
attribute caused Android to destroy the entire call task ~5s after the
user navigated away via task switch, killing the call.

Guard all teardown in onDestroy (signaling listeners, localStream,
foreground service, proximity sensor, broadcast receiver) so that
system-initiated destruction during task switching doesn't tear down
active call resources. The foreground service keeps the process alive.

Simplify onTopResumedActivityChanged to only enter PIP on API 29-30.
On API 31+, auto-enter handles swipe-up; onUserLeaveHint moves the task
to back as a safety net for task switching.

Signed-off-by: Tarek Loubani <tarek@tarek.org>

tareko commented:
Hello friends. After several weeks of using the app almost exclusively for my various calls, I believe this issue is now resolved with the pull above. Ultimately, there was no way to maintain PIP in the four main cases I ran into:
1. Fast swipe up on navigation/gesture bar
2. Slow swipe up on navigation/gesture bar
3. Swipe right on gesture bar (task switch)
4. Back gesture

After very frustrating attempts in which 2-3/4 of these would work, I reviewed other apps that have backgrounding features, and ultimately concluded that no modern app currently has all of these features. Specifically, I looked at Whatsapp (swipe up and back work to get into PIP, but not swipe right to task switch).

That left the matter of swipe-right resulting in the call getting disconnected, which is why lots of the "did the user intend to hang up" logic came into play. I looked at the code holistically again, and I believe the code is as clean as I know how to make it. I used a mix of Qwen, GLM, Codex, and Claude to debug the issues.

Hopefully the code can be reviewed and incorporated, as I would rather not have all of our users using our organization's fork. Thank you all for a great app, and hopefully we can get more done soon!

backfield79 commented:
> Hello friends. After several weeks of using the app almost exclusively for my various calls, I believe this issue is now resolved with the pull above. Ultimately, there was no way to maintain PIP in the four main cases I ran into:
> 1. Fast swipe up on navigation/gesture bar
> 2. Slow swipe up on navigation/gesture bar
> 3. Swipe right on gesture bar (task switch)
> 4. Back gesture
> 
> After very frustrating attempts in which 2-3/4 of these would work, I reviewed other apps that have backgrounding features, and ultimately concluded that no modern app currently has all of these features. Specifically, I looked at Whatsapp (swipe up and back work to get into PIP, but not swipe right to task switch).
> 
> That left the matter of swipe-right resulting in the call getting disconnected, which is why lots of the "did the user intend to hang up" logic came into play. I looked at the code holistically again, and I believe the code is as clean as I know how to make it. I used a mix of Qwen, GLM, Codex, and Claude to debug the issues.
> 
> Hopefully the code can be reviewed and incorporated, as I would rather not have all of our users using our organization's fork. Thank you all for a great app, and hopefully we can get more done soon!

Whish this PR be merged soon😎

abdusalamApps commented:
> Hello friends. After several weeks of using the app almost exclusively for my various calls, I believe this issue is now resolved with the pull above. Ultimately, there was no way to maintain PIP in the four main cases I ran into:
> 
>     1. Fast swipe up on navigation/gesture bar
> 
>     2. Slow swipe up on navigation/gesture bar
> 
>     3. Swipe right on gesture bar (task switch)
> 
>     4. Back gesture
> 
> 
> After very frustrating attempts in which 2-3/4 of these would work, I reviewed other apps that have backgrounding features, and ultimately concluded that no modern app currently has all of these features. Specifically, I looked at Whatsapp (swipe up and back work to get into PIP, but not swipe right to task switch).
> 
> That left the matter of swipe-right resulting in the call getting disconnected, which is why lots of the "did the user intend to hang up" logic came into play. I looked at the code holistically again, and I believe the code is as clean as I know how to make it. I used a mix of Qwen, GLM, Codex, and Claude to debug the issues.
> 
> Hopefully the code can be reviewed and incorporated, as I would rather not have all of our users using our organization's fork. Thank you all for a great app, and hopefully we can get more done soon!

Thank you for the great work. This is a much needed feature. Jazakallahu Khayran. Inshallah the situation in Gaza gets better soon. Our duas and support are with you daily ❤️

Thank you for all Nextcloud contributors too!

Commit:
feat(call): use Notification.CallStyle for ongoing call notification

Use Android CallStyle notification (API 31+) to show green status bar
chip with call duration timer, matching the native phone app experience.
Falls back to standard notification on older API levels.

The notification is updated every second via startForeground() to keep
the call duration accurate, using callStartTime from
ApplicationWideCurrentRoomHolder.

Signed-off-by: Tarek Loubani <tarek@tarek.org>

mahibi commented:
Thank you to keep up the work :+1: 
Please let us know if you would like to resolve the latest merge conflicts.

Please also resolve the DCO warning by following the instruction on https://github.com/nextcloud/talk-android/pull/5660/checks?check_run_id=70520434780

Commit:
Fix call notification not appearing immediately and missing on subsequent calls

Start foreground service at the beginning of prepareCall() before heavy
initialization, stop it in hangup() and unconditionally in onDestroy(),
cancel stale periodic handlers in onStartCommand(), and reset
callStartTime between calls to prevent state leakage.

Signed-off-by: Tarek Loubani <tarek@tarek.org>

Commit:
style: fix Codacy issues (line length, unused imports, trailing whitespace, generic catch)

- Break long log lines to respect 120 char limit
- Remove unused imports (LiveData, assertFalse)
- Remove trailing whitespace
- Merge duplicate test to reduce class function count below threshold
- Catch IllegalArgumentException instead of generic Exception
- Ensure EndCallReceiver.kt ends with newline

Signed-off-by: Tarek Loubani <tarek@tarek.org>

Commit:
Remove trailing space

Signed-off-by: Tarek Loubani <tarek@tarek.org>

Commit:
Remove trailing spaces

Signed-off-by: Tarek Loubani <tarek@tarek.org>

Commit:
Correctly hang up from notification in Android 12+

Signed-off-by: Tarek Loubani <tarek@tarek.org>

Commit:
fix: remove stray closing brace in ChatActivityLeaveRoomLifecycleTest

Signed-off-by: Tarek Loubani <tarek@tarek.org>

Commit:
style: remove trailing whitespace in CallActivity and CallForegroundService

Signed-off-by: Tarek Loubani <tarek@tarek.org>

Commit:
style: remove trailing whitespace in CallActivity and CallForegroundService

Signed-off-by: Tarek Loubani <tarek@tarek.org>

Commit:
style: remove trailing whitespace in CallActivity

Signed-off-by: Tarek Loubani <tarek@tarek.org>

tareko commented:
I believe the code is now in good enough shape for final review and pull.

mahibi commented:
> I believe the code is now in good enough shape for final review and pull.

thanks keeping it up. we will review soon

Commit:
chore: restore verification-metadata.xml from master

Signed-off-by: Tarek Loubani <tarek@tarek.org>

tareko commented:
Poking here for review.

mahibi commented:
> Poking here for review.

noted. sorry for the delay.

mahibi commented:
@tareko as the branch is behind the upstream master quite a bit it makes sense to pull the latest upstream changes and rebase. github already unveils there is a merge conflict with the CallActivity.

Sorry again for the huge delay, i know this is not satisfying. It's currently challenging to keep up with the amount of PRs.

Regardless if you are willing to keep up the work until this is merged: The best solution would be to move this PR to the https://github.com/nextcloud/talk-android repository. You should already have the permissions granted to create PRs in our repo. In the event that further delays occur, Nextcloud employees could also rebase on latest master and take over anytime if you like.

Thanks again for contributing!

backfield79 commented:
Hi! Is there a hope to be released in Talk 24?

mahibi commented:
> Hi! Is there a hope to be released in Talk 24?

i will try! But when i'm not confinced about the stability i might move it to 24.1.0

backfield79 commented:
> > Hi! Is there a hope to be released in Talk 24?
> 
> i will try! But when i'm not confinced about the stability i might move it to 24.1.0

Thank you so much! Waiting for the release)

Commit:
fix(build): add missing dependency verification checksums and trusted keys

Adds Gradle distribution key (1BD97A6A) for gradle-9.5.1-src.zip and
regenerates verification metadata for org.jetbrains.kotlin.plugin.compose
2.3.21 and other updated dependencies.

Also removes stray closing brace in ChatActivity.

AI-assistant: OpenCode (deepseek-v4-pro)
Signed-off-by: Tarek Loubani <tarek@tarek.org>

tareko commented:
OK, this should now be properly rebased and ready. I also tested it locally again and everything remains functional. Can we please get this figured out before we have to rebase it yet again?

mahibi commented:
Thank you @tareko 
It's on my list for monday!
So won't be included in 24.0.0 but i will make sure it will follow soon.

tareko commented:
> Thank you @tareko 
> It's on my list for monday!
> So won't be included in 24.0.0 but i will make sure it will follow soon.

I'll make a note to check there are no blocking changes on Sunday.

mahibi commented:
Hi @tareko i gave it another try today and came across some bugs:

- android user not hearable after joining. android user has to disable and enable the mic, after this it works

- when accepting via call notification, it can happen that the call is started in a second instance which is hidden (see in recents apps)

- user avatar of other participant is phone icon instead of the real avatar

- call notification is shown when opening a call. This is a problem since the last PR.
	- expected: only show the notification when call goes to background
 
- call screen content overlaps with status bar

- 1. bring the app to background while the call continues
- 2. open the app via the app icon
- 3. join some other chat
	- call hangs up on other side, but call service on android continues.
	- Expected "for now": show an info its not possible to join other chat


I will continue tomorrow and have a closer look.
Are you okay if i take over the PR? I would move it to the nextcloud repo and continue there until it's merged.

tareko commented:
Yes, please take over the PR. I don't think I can do much more with it, especially as I can't reproduce the bugs you've found. Thank you!

Out of curiosity, what device are you using?

mahibi commented:
> Yes, please take over the PR. I don't think I can do much more with it, especially as I can't reproduce the bugs you've found. Thank you!
> 
> Out of curiosity, what device are you using?

Ok thanks to let me know, i will take over and migrate the PR to the nextcloud repo. 
I tested this on Motorola Edge 50Pro with Android 16 but will do further testing with more devices / Android versions.

Thanks again for contributing @tareko and sorry again the review took so long.

mahibi commented:
i will close in favor of https://github.com/nextcloud/talk-android/pull/6334
